batches.go

```diff
@@ -10,52 +11,55 @@ const batchBasePath = "/batches"
 	Delete(string, bool) (*Response, error)
 }
 
 // Batch type
 type Batch struct {
-	ID                     string     `json:"id"`
-	State                  string     `json:"state,omitempty"`
-	Quantity               int32      `json:"quantity,omitempty"`
-	CreatedAt              *Timestamp `json:"created_at,omitempty"`
-	Href                   string     `json:"href,omitempty"`
-	Project                Href       `json:"project,omitempty"`
-	Facilities             []Facility `json:"facilities,omitempty"`
-	FacilityDiversityLevel int32      `json:"facility_diversity_level,omitempty"`
-	Devices                []Device   `json:"devices,omitempty"`
+	ID         string     `json:"id"`
+	State      string     `json:"state,omitempty"`
+	Quantity   int32      `json:"quantity,omitempty"`
+	CreatedAt  *Timestamp `json:"created_at,omitempty"`
+	Href       string     `json:"href,omitempty"`
+	Project    Href       `json:"project,omitempty"`
+	Facilities []Facility `json:"facilities,omitempty"`
+	Devices    []Device   `json:"devices,omitempty"`
 }
 
 //BatchesList represents collection of batches
 type batchesList struct {
 	Batches []Batch `json:"batches,omitempty"`
 }
 
-// InstanceBatchCreateRequest type used to create batch of device instances
-type InstanceBatchCreateRequest struct {
-	Batches []BatchInstance `json:"batches"`
+// BatchDeviceCreateRequest type used to create batch of device instances
+type BatchDeviceCreateRequest struct {
+	Batches []BatchCreateDevice `json:"batches"`
 }
 
-// BatchInstance type used to describe batch instances
-type BatchInstance struct {
-	Plan            string     `json:"plan"`
-	Hostname        string     `json:"hostname"`
-	Facility        string     `json:"facility"`
-	BillingCycle    string     `json:"billing_cycle"`
-	OperatingSystem string     `json:"operating_system"`
-	Quantity        int        `json:"quantity"`
-	Hostnames       []string   `json:"hostnames,omitempty"`
-	Description     string     `json:"description,omitempty"`
-	AlwaysPxe       bool       `json:"always_pxe,omitempty"`
-	Userdata        string     `json:"userdata,omitempty"`
-	Locked          bool       `json:"locked,omitempty"`
-	TerminationTime *Timestamp `json:"termination_time,omitempty"`
-	Tags            []string   `json:"tags,omitempty"`
-	ProjectSSHKeys  []string   `json:"project_ssh_keys,omitempty"`
-	UserSSHKeys     []string   `json:"user_ssh_keys,omitempty"`
-	Features        []string   `json:"features,omitempty"`
-	Customdata      string     `json:"customdata,omitempty"`
+// BatchCreateDevice type used to describe batch instances
+type BatchCreateDevice struct {
+	DeviceCreateRequest
+	Quantity               int32 `json:"quantity"`
+	FacilityDiversityLevel int32 `json:"facility_diversity_level,omitempty"`
+}
+
+// MarshalJSON custom marshalling to handle DeviceCreateRequest and additional fields
+func (bcd *BatchCreateDevice) MarshalJSON() ([]byte, error) {
```

Comment:
I agree with t0mk serializing should work with json.Marshal.

Reply:
@t0mk or @jasmingacic  could you please provide a sample of how to avoid the custom marshalling? I'm not sure how to accomplish easily, since the underlying json that gets sent to the API can take either a `[]string` or a `string` for `facility`, which I don't believe there is a good way to model in go, unless we override the marshalling to look at `Facilities` and apply the logic there.

I also rather not change `Facilities` from `string` -> `[]string` because that would introduce a breaking change